server/resp/src/main/java/org/infinispan/server/resp/commands/cluster/SHARDS.java

```diff
@@ -0,0 +1,247 @@
+ * command to the current topology members to retrieve specific data from the nodes.
+ *
+ * @link <a href="https://redis.io/commands/cluster-shards/">CLUSTER SHARDS</a>
+ * @since 15.0
+ */
+public class SHARDS extends RespCommand implements Resp3Command {
+
+   @GuardedBy("this")
+   private CompletionStage<CharSequence> lastExecution = null;
+
+   @GuardedBy("this")
+   private ConsistentHash lastAcceptedHash = null;
+
+   public SHARDS() {
+      super(2, 0, 0, 0);
+   }
+
+   @Override
+   public CompletionStage<RespRequestHandler> perform(Resp3Handler handler, ChannelHandlerContext ctx, List<byte[]> arguments) {
+      AdvancedCache<?, ?> respCache = handler.cache();
+      DistributionManager dm = respCache.getDistributionManager();
+      if (dm == null) {
+         RespErrorUtil.customError("This instance has cluster support disabled", handler.allocator());
+         return handler.myStage();
+      }
+
+      CacheTopology topology = dm.getCacheTopology();
+      ConsistentHash hash = topology.getCurrentCH();
+
+      if (hash == null) {
+         RespErrorUtil.customError("No consistent hash available", handler.allocator());
+         return handler.myStage();
+      }
+
+      synchronized (this) {
+         if (!hash.equals(lastAcceptedHash)) {
+            lastExecution = readShardsInformation(hash, SecurityActions.getClusterExecutor(respCache));
+            lastAcceptedHash = hash;
+         }
+      }
+      return handler.stageToReturn(lastExecution, ctx, ByteBufferUtils::stringToByteBuf);
+   }
+
+   private static CompletionStage<CharSequence> readShardsInformation(ConsistentHash hash, ClusterExecutor executor) {
+      Map<List<Address>, IntSet> segmentOwners = new HashMap<>();
+      for (int i = 0; i < hash.getNumSegments(); i++) {
+         segmentOwners.computeIfAbsent(hash.locateOwnersForSegment(i), ignore -> IntSets.mutableEmptySet())
+               .add(i);
+      }
+
+      return readNodeInformation(hash.getMembers(), executor)
+            .thenApply(information -> {
+               StringBuilder response = new StringBuilder();
+
+               // Number of elements.
+               response.append('*').append(segmentOwners.size()).append(CRLF);
+
+               for (Map.Entry<List<Address>, IntSet> entry : segmentOwners.entrySet()) {
+                  List<Address> addresses = entry.getKey();
+                  String leader = information.get(addresses.get(0));
+                  if (leader == null) {
+                     log.debugf("Not found information for leader: %s", addresses.get(0));
+                     String name = addresses.get(0).toString();
+                     StringBuilder sb = new StringBuilder();
+
+                     // A health of `loading` mean that the node will be available in the future for traffic.
+                     createNodeSerialized(sb, name, name, 0, "loading");
+                     leader = sb.toString();
```

Comment:
This doesn't look right, it will contain all the serialized info in RESP format but leader when it is not null only has the node name as far as I can see.

Reply:
When `leader` is not null the content is:

```text
*14
$2
id
$29
ClusteredOperationsTest-NodeA
$4
port
:55921
$2
ip
$9
127.0.0.1
$8
endpoint
$9
127.0.0.1
$18
replication-offset
:0
$6
health
$6
online
```

And when it is null:

```text
*14
$2
id
$29
ClusteredOperationsTest-NodeA
$4
port
:0
$2
ip
$29
ClusteredOperationsTest-NodeA
$8
endpoint
$29
ClusteredOperationsTest-NodeA
$18
replication-offset
:0
$6
health
$7
loading
```

Reply:
Oh, yes sorry, I missed the update to readLocalNodeInformation